src/Functions/URL/cutURLParameter.cpp

```diff
@@ -1,82 +1,183 @@
+            return col_res;
+        }
+        else
+            throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal column {} of argument of function {}",
+                arguments[0].column->getName(), getName());
+    }
+
+    static void cutURL(ColumnString::Chars & data, std::string pattern)
+    {
+        pattern += '=';
+        const char * param_str = pattern.c_str();
+        size_t param_len = pattern.size();
+
+        const char * url_begin = reinterpret_cast<const char *>(&data[0]);
+        const char * url_end = reinterpret_cast<const char *>(&data[data.size()-1]) - 1;
+        const char * begin_pos = url_begin;
+        const char * end_pos = begin_pos;
+
+        do
+        {
+            const char * query_string_begin = find_first_symbols<'?', '#'>(url_begin, url_end);
+            if (query_string_begin + 1 >= url_end)
+                break;
+
+            const char * pos = static_cast<const char *>(memmem(query_string_begin + 1, url_end - query_string_begin - 1, param_str, param_len));
+            if (pos == nullptr)
+                break;
+
+            if (pos[-1] != '?' && pos[-1] != '#' && pos[-1] != '&')
+            {
+                pos = nullptr;
+                break;
+            }
+
+            begin_pos = pos;
+            end_pos = begin_pos + param_len;
+
+            /// Skip the value.
+            while (*end_pos && *end_pos != '&' && *end_pos != '#')
+                ++end_pos;
+
+            /// Capture '&' before or after the parameter.
+            if (*end_pos == '&')
+                ++end_pos;
+            else if (begin_pos[-1] == '&')
+                --begin_pos;
+        } while (false);
+
+        data.erase(data.begin() + (begin_pos - url_begin), data.begin() + (end_pos - url_begin));
+    }
+
     static void vector(const ColumnString::Chars & data,
         const ColumnString::Offsets & offsets,
-        std::string pattern,
+        const ColumnConst * col_needle,
+        const ColumnArray * col_const_array,
         ColumnString::Chars & res_data, ColumnString::Offsets & res_offsets)
     {
         res_data.reserve(data.size());
         res_offsets.resize(offsets.size());
 
-        pattern += '=';
-        const char * param_str = pattern.c_str();
-        size_t param_len = pattern.size();
-
         size_t prev_offset = 0;
         size_t res_offset = 0;
 
+        ColumnString::Chars res_chars;
+
         for (size_t i = 0; i < offsets.size(); ++i)
         {
             size_t cur_offset = offsets[i];
 
-            const char * url_begin = reinterpret_cast<const char *>(&data[prev_offset]);
-            const char * url_end = reinterpret_cast<const char *>(&data[cur_offset]) - 1;
-            const char * begin_pos = url_begin;
-            const char * end_pos = begin_pos;
+            res_chars.clear();
+            res_chars.insert(res_chars.end(), data.begin()+prev_offset, data.begin()+cur_offset);
```

Comment:
Cosmetic: add spaces around +

Reply:
Maybe I don't find it right now, but we previously did not do this copy, right? Maybe the URL could be copied into the result column and manipulated in-place? (for arrays, cutURL() is called repeatedly but the URL only gets shorter with each iteration, which means one just needs to adjust the result offset).

Reply:
Yes, in previous version (before this PR) we did not do the copy. I think it's a good idea to do manipulations "in place" on the result column. I will try to implement it. I thought cutURL()  could be some kind of independent function. But probably the approach "without copying" is faster.

Reply:
@rschu1ze Please take a look. I implemented it in another way:
first allocate res_data and then cut it after each call of `cutURL`.

Reply:
Close because discuss it in another thread.